File: Server/WarshipServer/CommandManager.cpp
#include "CommandManager.h"

int CommandManager::serialize(const Command& c, char* buf)
{
	if (c.id == Command::NameSend)
	{
		return serializeStringCommand((StringCommand&)c, buf);
	}
	else
	{
		return serializeIntArrayCommand((IntArrayCommand&)c, buf);
	}
}


int CommandManager::serializeIntArrayCommand(const IntArrayCommand& c, char* buf)
{
	int neededLen = 2 + c.data.size();
	
	buf[0] = (unsigned char)c.id;
	buf[1] = (unsigned char)c.parameter;

	for (size_t i = 0; i < c.data.size(); i++)
	{
		buf[i + 2] = c.data.at(i);
	}

	return neededLen;
}


int CommandManager::serializeStringCommand(const StringCommand& c, char* buf)
{
	int neededLen = 2 + c.data.size();

	buf[0] = (unsigned char)c.id;
	buf[1] = (unsigned char)c.parameter;

	for (size_t i = 0; i < c.data.size(); i++)
	{
		buf[i + 2] = c.data.at(i);
	}

	return neededLen;
}


std::unique_ptr<Command> CommandManager::deserialize(const char * buf, const int len)
{
	Command::CommandID id = (Command::CommandID)buf[0];

	std::unique_ptr<Command> commandPtr;

	if (id == Command::NameSend)
	{
		//Deserialize string
		std::unique_ptr<StringCommand> stringCommandPtr = std::make_unique<StringCommand>();

		stringCommandPtr->id = id;
		stringCommandPtr->parameter = buf[1];

		std::ostringstream stream;
		
		for (int i = 2; i < len; i++)
		{
			stream << buf[i];
		}

		stringCommandPtr->data = stream.str();

		commandPtr = std::move(stringCommandPtr);
	}
	else
	{
		//Deserialize int array
		std::unique_ptr<IntArrayCommand> intArrayCommandPtr = std::make_unique <IntArrayCommand>();

		intArrayCommandPtr->id = id;
		intArrayCommandPtr->parameter = buf[1];

		for (int i = 2; i < len; i++)
		{
			intArrayCommandPtr->data.push_back(buf[i]);
		}

		commandPtr = std::move(intArrayCommandPtr);
	}


	return commandPtr;
}


void CommandManager::testSerialization()
{
	char buf[32];
	StringCommand c;
	c.id = Command::NameSend;
	c.parameter = 0;
	c.data = "Zykyflex";

	int len = serialize(c, buf);

	for(int i = 0; i < len; ++i)
	{
		std::cout << (int)buf[i] << " ";
	}

	std::cout << std::endl;

	std::unique_ptr<Command> cptr = deserialize(buf, len);
	displayCommand(cptr.get());

	IntArrayCommand ic;
	ic.id = Command::TargetGrid;
	ic.parameter = 2;
	
	for (size_t i = 0; i < 5; i++)
	{
		ic.data.push_back(i*2 + 2);
	}

	len = serialize(ic, buf);

	for (int i = 0; i < len; ++i)
	{
		std::cout << (int)buf[i] << " ";
	}
	std::cout << std::endl;

	cptr = deserialize(buf, len);
	displayCommand(cptr.get());
}

void CommandManager::displayCommand(const Command* c)
{
	if (c->id == Command::NameSend)
	{
		displayStringCommand((StringCommand*)c);
	}
	else
	{
		displayIntArrayCommand((IntArrayCommand*)c);
	}
}

std::string CommandManager::getStringCommandID(const Command::CommandID commandID)
{
	switch (commandID)
	{
		case Command::IDAttrib:
			return "UDAttrib";
		case Command::RegisterOther:
			return "RegisterOther";
		case Command::NameSend:
			return "NameSend";
		case Command::TargetGrid:
			return "TargetGrid";
		case Command::FireGrid:
			return "FireGrid";
		case Command::FireReady:
			return "FireReady";
		case Command::FireResult:
			return "FireResult";
		case Command::ShipSunk:
			return "ShipSunk";
		case Command::PlaceShip:
			return "PlaceShip";
		case Command::GameStarts:
			return "GameStarts";
		case Command::EndGame:
			return "EndGame";
		default:
			return "Forgot CommandManager-getStringCommandID " + commandID;


	}
}

void CommandManager::displayStringCommand(const StringCommand* c)
{
	std::ostringstream stream;
	stream << "(" << c->socketID << ") " << getStringCommandID(c->id) << " " << c->parameter << " : " << c->data << std::endl;

	std::cout << stream.str();
}

void CommandManager::displayIntArrayCommand(const IntArrayCommand* c)
{
	std::ostringstream stream;
	stream << "(" << c->socketID << ") " << getStringCommandID(c->id) << " " << c->parameter << " : ";

	for (size_t i = 0; i < c->data.size(); i++)
	{
		stream << c->data.at(i) << " ";
	}

	std::cout << stream.str() << std::endl;
}

File: Client/WarshipClient/Assets/Scripts/Network/CommandManager.cs
using System.Collections;
using System.Collections.Generic;
using System.Text;
using UnityEngine;

public class CommandManager : MonoBehaviour
{
    public enum CommandID { IDAttrib, RegisterOther, NameSend, TargetGrid, FireGrid, FireResult, ShipSunk, EndGame, PlaceShip, GameStarts }


    public void testSerialization()
    {
        //TEST
        StringCommand nameCommand = new StringCommand();
        nameCommand.id = CommandID.NameSend;
        nameCommand.param = 0;
        nameCommand.data = "Zykyflex";

        displayCommand(nameCommand);

        byte[] buf = new byte[255];

        int len = serializeCommand(nameCommand, buf);

        Command deserialized = deserializeCommand(buf, len);

        displayCommand(deserialized);

        IntArrayCommand intCommand = new IntArrayCommand();
        intCommand.id = CommandID.TargetGrid;
        intCommand.param = 1;
        intCommand.data =  new int[]{ 1,2,3,4,5,6,7,8,9};

        displayCommand(intCommand);

        len = serializeCommand(intCommand, buf);
        deserialized = deserializeCommand(buf, len);

        displayCommand(deserialized);


    }

    public Command deserializeCommand(byte[] rawCommand, int len)
    {
        //Command Code
        CommandID cid = (CommandID)rawCommand[0];
        int parameter = rawCommand[1];

        if(cid == CommandID.NameSend)
        {
            //String command
            StringCommand command = new StringCommand();
            command.id = cid;
            command.param = parameter;

            StringBuilder builder = new StringBuilder();

            for (int i = 2; i < len; i++)
            {
                builder.Append((char)rawCommand[i]);
            }

            command.data = builder.ToString();

            return command;
        }
        else
        {
            //int[] command
            IntArrayCommand command = new IntArrayCommand();
            command.id = cid;
            command.param = parameter;

            int[] data = new int[len - 2];
            for (int i = 2; i < len; i++)
            {
                data[i - 2] = rawCommand[i];
            }

            command.data = data;

            return command;
        }

    }

    public int serializeCommand(Command command, byte[] buf)
    {
        if(command.id == CommandID.NameSend)
        {
            return serializeCommand((StringCommand)command, buf);
        }
        else
        {
            return serializeCommand((IntArrayCommand)command, buf);
        }
    }

    public int serializeCommand(IntArrayCommand command, byte[] buf)
    {
        buf[0] = (byte)command.id;
        buf[1] = (byte)command.param;

        int i;
        for (i = 0; i < command.data.Length; i++)
        {
            byte b;
            if(command.data[i] > 255)
            {
                Debug.LogWarning("Serializing int greater than 255 to 255.");
                b = 255;
            }
            else
            {
                b = (byte)command.data[i];
            }
            buf[i+2] = b;
        }

        return i+2;
    }

    public int serializeCommand(StringCommand command, byte[] buf)
    {
        buf[0] = (byte)command.id;
        buf[1] = (byte)command.param;

        int i;
        for (i = 0; i < command.data.Length; i++)
        {
            buf[i+2] = (byte)command.data[i];
        }

        return i + 2;
    }

    public void displayCommand(Command c)
    {
        if(c.id == CommandID.NameSend)
        {
            displayStringCommand((StringCommand)c);
        }
        else
        {
            displayIntArrayCommand((IntArrayCommand)c);
        }
    }

    public void displayStringCommand(StringCommand c)
    {
        Debug.Log(c.id + " " + c.param + ": " + c.data);
    }

    public void displayIntArrayCommand(IntArrayCommand c)
    {
        StringBuilder sb = new StringBuilder();
        sb.Append(c.id);
        sb.Append(" ");
        sb.Append(c.param);
        sb.Append(": ");

        for (int i = 0; i < c.data.Length; i++)
        {
            sb.Append(c.data[i]);
            sb.Append(" ");
        }

        Debug.Log(sb.ToString());
    }
}

public class Command
{
    public CommandManager.CommandID id;
    public int param;
}

public class StringCommand : Command
{
    public string data;
}

public class IntArrayCommand : Command
{
    public int[] data;
}
